File: sight/java/org/jboss/sight/Volume.java
/*
 *  SIGHT - System information gathering hybrid tool
 *
 *  Copyright(c) 2007 Red Hat Middleware, LLC,
 *  and individual contributors as indicated by the @authors tag.
 *  See the copyright.txt in the distribution for a
 *  full listing of individual contributors.
 *
 *  This library is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU Lesser General Public
 *  License as published by the Free Software Foundation; either
 *  version 2 of the License, or (at your option) any later version.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public
 *  License along with this library in the file COPYING.LIB;
 *  if not, write to the Free Software Foundation, Inc.,
 *  59 Temple Place - Suite 330, Boston, MA 02111-1307, USA
 *
 */

package org.jboss.sight;

import java.util.Iterator;
import java.util.EnumSet;

/**
 * Volume information
 *
 * @author Mladen Turk
 *
 */

public class Volume extends NativeObject
{

    private static native long      enum0(long pool);
    private static native int       enum1(long handle);
    private static native int       enum2(Volume thiz, long handle);
    private static native int       enum3(Volume thiz, long handle);
    private static native int       enum4(long handle);

    private Volume()
    {
        super(0);
    }

    protected void onDestroy()
    {
        // Nothing
    }

    private void setFSType(int type)
    {
        Type = FileSystemType.valueOf(type);
    }


    private void setFlags(int flags)
    {
        Flags = FileSystemFlags.valueOf(flags);
    }

    private void setDType(int flags)
    {
        DriveType = DriveType.valueOf(flags);
    }


    /**
     * Returns the array of all Volumes
     */
    public static VolumeIterator getVolumes()
        throws OutOfMemoryError, OperatingSystemException
    {
        Volume[] array;
        long handle = enum0(0);
        try {
            int cnt = enum1(handle);
            array = new Volume[cnt];
            for (int i = 0; i < cnt; i++) {
                array[i] = new Volume();
                enum2(array[i], handle);
                int numberOfMounts = array[i].NumberOfMountPoints;
                for (int j = 0; j < numberOfMounts; j++) {
                    Volume mountPoint = new Volume();
                    int rv = enum3(mountPoint, handle);
                    if (rv != 0)
                        array[++i] = mountPoint;
                }
            }
            return new VolumeIterator(array);
        }
        finally {
            // Do not leak memory
            enum4(handle);
        }
    }

    /**
     * Name of the volume.
     */
    public String               Name;

    /**
     * Description for the volume.
     */
    public String               Description;

    /**
     * Volume Mount Point.
     * <BR/>
     * This is directory the volume is mouted on,
     * eg <code>C:\</code> on windows or
     * <code>/home</code> on unixes.
     */
    public String               MountPoint;

    /**
     * File system type.
     */
    public FileSystemType       Type;

    /**
     * Determines whether a disk drive is a removable, fixed, CD-ROM,
     * RAM disk, or network drive.
     */
    public DriveType            DriveType;

    /**
     * The number of sectors per cluster.
     */
    public int                  SectorsPerCluster;

    /**
     * The number of bytes per sector.
     */
    public int                  BytesPerSector;

    /**
     * Total number of free bytes on a disk that are available to the
     * user who is associated with the call.
     */
    public long                 FreeBytesAvailable;

    /**
     * Total number of free bytes on a disk.
     */
    public long                 TotalNumberOfBytes;

    /**
     * TotalNumberOfFreeBytes.
     */
    public long                 TotalNumberOfFreeBytes;

    /**
     * File system flags.
     */
    public EnumSet<FileSystemFlags> Flags;

    /**
     * Number of mounted volume points.
     */
    public int                  NumberOfMountPoints;
}

File: sight/native/os/stubs/volume.c
/*
 *  SIGHT - System information gathering hybrid tool
 *
 *  Copyright(c) 2007 Red Hat Middleware, LLC,
 *  and individual contributors as indicated by the @authors tag.
 *  See the copyright.txt in the distribution for a
 *  full listing of individual contributors.
 *
 *  This library is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU Lesser General Public
 *  License as published by the Free Software Foundation; either
 *  version 2 of the License, or (at your option) any later version.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public
 *  License along with this library in the file COPYING.LIB;
 *  if not, write to the Free Software Foundation, Inc.,
 *  59 Temple Place - Suite 330, Boston, MA 02111-1307, USA
 *
 * @author Mladen Turk
 *
 */

#include "sight.h"
#include "sight_local.h"
#include "sight_types.h"

#define SIGHT_WANT_LATE_DLL
#include "sight_private.h"

/*
 * Network adapter implementation
 */

J_DECLARE_CLAZZ = {
    NULL,
    NULL,
    SIGHT_CLASS_PATH "Volume"
};

J_DECLARE_F_ID(0000) = {
    NULL,
    "Name",
    "Ljava/lang/String;"
};

J_DECLARE_F_ID(0001) = {
    NULL,
    "Description",
    "Ljava/lang/String;"
};

J_DECLARE_F_ID(0002) = {
    NULL,
    "MountPoint",
    "Ljava/lang/String;"
};

J_DECLARE_F_ID(0003) = {
    NULL,
    "SectorsPerCluster",
    "I"
};

J_DECLARE_F_ID(0004) = {
    NULL,
    "BytesPerSector",
    "I"
};

J_DECLARE_F_ID(0005) = {
    NULL,
    "FreeBytesAvailable",
    "J"
};

J_DECLARE_F_ID(0006) = {
    NULL,
    "TotalNumberOfBytes",
    "J"
};

J_DECLARE_F_ID(0007) = {
    NULL,
    "TotalNumberOfFreeBytes",
    "J"
};

J_DECLARE_F_ID(0008) = {
    NULL,
    "NumberOfMountPoints",
    "I"
};

J_DECLARE_M_ID(0000) = {
    NULL,
    "setFSType",
    "(I)V"
};

J_DECLARE_M_ID(0001) = {
    NULL,
    "setFlags",
    "(I)V"
};

J_DECLARE_M_ID(0002) = {
    NULL,
    "setDType",
    "(I)V"
};


SIGHT_CLASS_LDEF(Volume)
{
    if (sight_load_class(_E, &_clazzn))
        return 1;
    J_LOAD_IFIELD(0000);
    J_LOAD_IFIELD(0001);
    J_LOAD_IFIELD(0002);
    J_LOAD_IFIELD(0003);
    J_LOAD_IFIELD(0004);
    J_LOAD_IFIELD(0005);
    J_LOAD_IFIELD(0006);
    J_LOAD_IFIELD(0007);
    J_LOAD_IFIELD(0008);
    J_LOAD_METHOD(0000);
    J_LOAD_METHOD(0001);
    J_LOAD_METHOD(0002);

    return 0;
}

SIGHT_CLASS_UDEF(Volume)
{
    sight_unload_class(_E, &_clazzn);
}

/* Initialize volume enumeration */
SIGHT_EXPORT_DECLARE(jlong, Volume, enum0)(SIGHT_STDARGS,
                                           jlong pool)
{
    return 0;
}

SIGHT_EXPORT_DECLARE(jint, Volume, enum1)(SIGHT_STDARGS,
                                          jlong handle)
{
    return 0;
}

SIGHT_EXPORT_DECLARE(void, Volume, enum2)(SIGHT_STDARGS,
                                          jobject thiz,
                                          jlong handle)
{
}

SIGHT_EXPORT_DECLARE(jint, Volume, enum3)(SIGHT_STDARGS,
                                          jobject thiz,
                                          jlong handle)
{
    return 0;
}

/* Close volume enumeration */
SIGHT_EXPORT_DECLARE(void, Volume, enum4)(SIGHT_STDARGS,
                                          jlong handle)
{
}
